File: spark/src/main/scala/geotrellis/spark/tiling/MapKeyTransform.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.spark.tiling

import geotrellis.spark._
import geotrellis.raster._
import geotrellis.raster.rasterize.{Rasterizer, Callback}
import geotrellis.vector._
import geotrellis.proj4._
import geotrellis.util._

import java.util.concurrent.ConcurrentHashMap
import scala.collection.JavaConverters._

object MapKeyTransform {
  def apply(crs: CRS, level: LayoutLevel): MapKeyTransform =
    apply(crs.worldExtent, level.layout.layoutCols, level.layout.layoutRows)

  def apply(crs: CRS, layoutDimensions: (Int, Int)): MapKeyTransform =
    apply(crs.worldExtent, layoutDimensions)

  def apply(crs: CRS, layoutCols: Int, layoutRows: Int): MapKeyTransform =
    apply(crs.worldExtent, layoutCols, layoutRows)

  def apply(extent: Extent, layoutDimensions: (Int, Int)): MapKeyTransform =
    apply(extent, layoutDimensions._1, layoutDimensions._2)

  def apply(extent: Extent, layoutCols: Int, layoutRows: Int): MapKeyTransform =
    new MapKeyTransform(extent, layoutCols, layoutRows)
}

/**
  * Transforms between geographic map coordinates and spatial keys.
  * Since geographic point can only be mapped to a grid tile that contains that point,
  * transformation from Extent to GridBounds to Extent will likely not
  * produce the original geographic extent, but a larger one.
  */
class MapKeyTransform(val extent: Extent, val layoutCols: Int, val layoutRows: Int) extends Serializable {
  lazy val tileWidth: Double = extent.width / layoutCols
  lazy val tileHeight: Double = extent.height / layoutRows

  def extentToBounds(otherExtent: Extent): GridBounds = apply(otherExtent)

  def apply(otherExtent: Extent): GridBounds = {
    val SpatialKey(colMin, rowMin) = apply(otherExtent.xmin, otherExtent.ymax)

    // For calculating GridBounds, the extent parameter is considered
    // inclusive on it's north and west borders, and execlusive on
    // it's east and south borders.
    // If the Extent has xmin == xmax and/or ymin == ymax, then consider
    // those zero length dimensions to represent the west and/or east
    // borders (so they are inclusive). In this case, the tiles returned
    // will be south and/or east of the line or point.
    val colMax = {
      val d = (otherExtent.xmax - extent.xmin) / (extent.width / layoutCols)

      if(d == math.floor(d) && d != colMin) { d.toInt - 1 }
      else { d.toInt }
    }

    val rowMax = {
      val d = (extent.ymax - otherExtent.ymin) / (extent.height / layoutRows)

      if(d == math.floor(d) && d != rowMin) { d.toInt - 1 }
      else { d.toInt }
    }

    GridBounds(colMin, rowMin, colMax, rowMax)
  }

  def boundsToExtent(gridBounds: GridBounds): Extent = apply(gridBounds)

  def apply(gridBounds: GridBounds): Extent = {
    val e1 = apply(gridBounds.colMin, gridBounds.rowMin)
    val e2 = apply(gridBounds.colMax, gridBounds.rowMax)
    e1.expandToInclude(e2)
  }

  /** Fetch the [[SpatialKey]] that corresponds to some coordinates in some CRS on the Earth. */
  def pointToKey(p: Point): SpatialKey = apply(p)

  def apply(p: Point): SpatialKey = apply(p.x, p.y)

  /** Fetch the [[SpatialKey]] that corresponds to some coordinates in some CRS on the Earth. */
  def pointToKey(x: Double, y: Double): SpatialKey = apply(x, y)

  def apply(x: Double, y: Double): SpatialKey = {
    val tcol =
      ((x - extent.xmin) / extent.width) * layoutCols

    val trow =
      ((extent.ymax - y) / extent.height) * layoutRows

    (tcol.floor.toInt, trow.floor.toInt)
  }

  @deprecated("Use this instead: MapKeyTransform.keyToExtent(key.getComponent[SpatialKey])", "1.2")
  def apply[K: SpatialComponent](key: K): Extent = apply(key.getComponent[SpatialKey])

  /** Get the [[Extent]] corresponding to a [[SpatialKey]] in some zoom level. */
  def keyToExtent(key: SpatialKey): Extent = apply(key)

  def apply(key: SpatialKey): Extent = apply(key.col, key.row)

  /** 'col' and 'row' correspond to a [[SpatialKey]] column and row in some grid. */
  def keyToExtent(col: Int, row: Int): Extent = apply(col, row)

  def apply(col: Int, row: Int): Extent =
    Extent(
      extent.xmin + col * tileWidth,
      extent.ymax - (row + 1) * tileHeight,
      extent.xmin + (col + 1) * tileWidth,
      extent.ymax - row * tileHeight
    )

  def multiLineToKeys(multiLine: MultiLine): Iterator[SpatialKey] = {
    val extent = multiLine.envelope
    val bounds: GridBounds = extentToBounds(extent)
    val options = Rasterizer.Options(includePartial=true, sampleType=PixelIsArea)

    val boundsExtent: Extent = boundsToExtent(bounds)
    val rasterExtent = RasterExtent(boundsExtent, bounds.width, bounds.height)

    /*
     * Use the Rasterizer to construct  a list of tiles which meet
     * the  query polygon.   That list  of tiles  is stored  as an
     * array of  tuples which  is then  mapped-over to  produce an
     * array of KeyBounds.
     */
    val tiles = new ConcurrentHashMap[(Int,Int), Unit]
    val fn = new Callback {
      def apply(col : Int, row : Int): Unit = {
        val tile : (Int, Int) = (bounds.colMin + col, bounds.rowMin + row)
        tiles.put(tile, Unit)
      }
    }

    multiLine.foreach(rasterExtent, options)(fn)
    tiles.keys.asScala.map { case (col, row) => SpatialKey(col, row) }
  }

  def multiPolygonToKeys(multiPolygon: MultiPolygon): Iterator[SpatialKey] = {
    val extent = multiPolygon.envelope
    val bounds: GridBounds = extentToBounds(extent)
    val options = Rasterizer.Options(includePartial=true, sampleType=PixelIsArea)
    val boundsExtent: Extent = boundsToExtent(bounds)
    val rasterExtent = RasterExtent(boundsExtent, bounds.width, bounds.height)

    /*
     * Use the Rasterizer to construct  a list of tiles which meet
     * the  query polygon.   That list  of tiles  is stored  as an
     * array of  tuples which  is then  mapped-over to  produce an
     * array of KeyBounds.
     */
    val tiles = new ConcurrentHashMap[(Int,Int), Unit]
    val fn = new Callback {
      def apply(col : Int, row : Int): Unit = {
        val tile : (Int, Int) = (bounds.colMin + col, bounds.rowMin + row)
        tiles.put(tile, Unit)
      }
    }

    multiPolygon.foreach(rasterExtent, options)(fn)
    tiles.keys.asScala.map { case (col, row) => SpatialKey(col, row) }
  }
}

File: geotrellis-python/geotrellis/spark/tiling/MapKeyTransform.py
from geotrellis.raster.GridBounds import GridBounds
from geotrellis.spark.SpatialKey import SpatialKey
from geotrellis.spark.tiling.package_scala import worldExtent
from geotrellis.vector.Extent import Extent
from geotrellis.proj4.CRS import CRS
from geotrellis.spark.tiling.LayoutScheme import LayoutLevel

class MapKeyTransform(object):
    def __init__(self, first, second, third = None):
        extent, layoutCols, layoutRows = _get_params(first, second, third)
        self.extent = extent
        self.layoutCols = layoutCols
        self.layoutRows = layoutRows
        self._tileWidth = None
        self._tileHeight = None

    @property
    def tileWidth(self):
        if self._tileWidth is not None:
            return self._tileWidth
        self._tileWidth = self.extent.width / layoutCols
        return self._tileWidth

    @property
    def tileHeight(self):
        if self._tileHeight is not None:
            return self._tileHeight
        self._tileHeight = self.extent.height / layoutRows
        return self._tileHeight

    def __call__(self, first, second = None):
        if second is not None and isinstance(first, float) and isinstance(second, float):
            x, y = first, second
            tcol = ((x - self.extent.xmin) / self.extent.width) * layoutCols
            trow = ((self.extent.ymax - y) / self.extent.height) * layoutRows
            return (int(tcol), int(trow))
        elif second is not None:
            col, row = first, second
            return Extent(
                    self.extent.xmin + col * self.tileWidth,
                    self.extent.ymax - (row+1) * self.tileHeight,
                    self.extent.xmin + (col+1) * self.tileWidth,
                    self.extent.ymax - row * self.tileHeight)
        elif isinstance(first, SpatialKey):
            key = first
            self(key.col, key.row)
        elif isinstance(first, GridBounds):
            gridBounds = first
            e1 = self(gridBounds.colMin, gridBounds.rowMin)
            e2 = self(gridBounds.colMax, gridBounds.rowMax)
            return e1.expandToInclude(e2)
        elif isinstance(first, Extent):
            otherExtent = first
            key = self(otherExtent.xmin, otherExtent.ymax)
            colMin = key.col
            rowMin = key.row
            d = (otherExtent.xmax - self.extent.xmin) / (self.extent.width / self.layoutCols)
            if d == math.floor(d) and d != colMin:
                colMax = int(d) - 1
            else:
                colMax = int(d)

            d = (self.extent.ymax - otherExtent.ymin) / (self.extent.height / self.layoutRows)
            if d == math.floor(d) and d != rowMin:
                rowMax = int(d) - 1
            else:
                rowMax = int(d)

            return GridBounds(colMin, rowMin, colMax, rowMax)
        else:
            raise Exception("wrong params: ({first}, {second})".format(
                first = first,
                second = second))

def _get_params(first, second, third):
    if isinstance(first, CRS) and isinstance(second, LayoutLevel) and third is None:
        crs = first
        level = second
        return worldExtent(crs), level.layout.layoutCols, level.layout.layoutRows
    elif isinstance(first, CRS) and isinstance(second, tuple) and third is None:
        crs = first
        layoutDimensions = second
        return worldExtent(crs), layoutDimensions[0], layoutDimensions[1]
    elif isinstance(first, CRS) and isinstance(second, int) and isinstance(third, int):
        crs = first
        layoutCols = second
        layoutRows = third
        return worldExtent(crs), layoutCols, layoutRows
    elif isinstance(first, Extent) and isinstance(second, tuple) and third is None:
        extent = first
        layoutDimensions = second
        return extent, layoutDimensions[0], layoutDimensions[1]
    elif isinstance(first, Extent) and isinstance(second, int) and isinstance(third, int):
        return first, second, third
    else:
        raise Exception("wrong params ({one}, {two}, {three})".format(
            one = first, two = second, three = third))
